File: Project-source/forge-verifier/csharp/src/HeartbeatClient.cs
// Heartbeat client — POSTs to /api/v1/licenses/{id}/heartbeat with HMAC-SHA256
// body signature. Wire protocol matches Go / Python SDKs byte-for-byte.

using System.Net;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace YourCo.Forge.Verifier;

public sealed class HeartbeatClient
{
    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private readonly string _baseUrl;
    private readonly string _apiKey;
    private readonly string _userAgent;
    private readonly HttpClient _http;

    public HeartbeatClient(string baseUrl, string apiKey, string userAgent, HttpClient? http = null)
    {
        _baseUrl = baseUrl;
        _apiKey = apiKey;
        _userAgent = userAgent;
        _http = http ?? new HttpClient();
    }

    public async Task<HeartbeatResponse> SendAsync(string licenseId, string fingerprint, CancellationToken ct = default)
    {
        var body = new HeartbeatRequest
        {
            LicenseId = licenseId,
            Fingerprint = fingerprint,
            VerifierVersion = _userAgent,
            Nonce = RandomNonceHex(),
            ReportedAt = DateTimeOffset.UtcNow.ToString("o"),
        };
        var raw = JsonSerializer.SerializeToUtf8Bytes(body, JsonOpts);
        var sig = HmacHex(_apiKey, raw);

        using var req = new HttpRequestMessage(HttpMethod.Post,
            $"{_baseUrl}/licenses/{licenseId}/heartbeat")
        {
            Content = new ByteArrayContent(raw)
            {
                Headers = { { "Content-Type", "application/json" } },
            },
        };
        req.Headers.TryAddWithoutValidation("X-Forge-API-Key", _apiKey);
        req.Headers.TryAddWithoutValidation("X-Forge-Signature", $"sha256={sig}");
        if (!string.IsNullOrEmpty(_userAgent))
        {
            req.Headers.TryAddWithoutValidation("User-Agent", _userAgent);
        }

        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        switch ((int)resp.StatusCode)
        {
            case 401: throw new InvalidOperationException("heartbeat: invalid api key (401)");
            case 429: throw new InvalidOperationException("heartbeat: rate limited (429)");
            case >= 500: throw new InvalidOperationException($"heartbeat: server error {(int)resp.StatusCode}");
        }
        var stream = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        return (await JsonSerializer.DeserializeAsync<HeartbeatResponse>(
            stream, JsonOpts, ct).ConfigureAwait(false))!;
    }

    internal static string HmacHex(string key, byte[] body)
    {
        using var mac = new HMACSHA256(Encoding.UTF8.GetBytes(key));
        return Convert.ToHexString(mac.ComputeHash(body)).ToLowerInvariant();
    }

    private static string RandomNonceHex()
    {
        Span<byte> b = stackalloc byte[16];
        RandomNumberGenerator.Fill(b);
        return Convert.ToHexString(b).ToLowerInvariant();
    }
}

public sealed class HeartbeatRequest
{
    [JsonPropertyName("license_id")] public string LicenseId { get; set; } = "";
    [JsonPropertyName("fingerprint")] public string Fingerprint { get; set; } = "";
    [JsonPropertyName("verifier_version")] public string VerifierVersion { get; set; } = "";
    [JsonPropertyName("nonce")] public string Nonce { get; set; } = "";
    [JsonPropertyName("reported_at")] public string ReportedAt { get; set; } = "";
}

public sealed class HeartbeatResponse
{
    [JsonPropertyName("ok")] public bool Ok { get; set; }
    [JsonPropertyName("license_status")] public string LicenseStatus { get; set; } = "";
    [JsonPropertyName("multi_env_anomaly")] public bool MultiEnvAnomaly { get; set; }
    [JsonPropertyName("next_heartbeat_after_seconds")] public int NextHeartbeatAfterSeconds { get; set; }
    [JsonPropertyName("reason")] public string? Reason { get; set; }
}

File: Project-source/forge-verifier/java/src/main/java/com/yourco/forge/verifier/HeartbeatClient.java
package com.yourco.forge.verifier;

import com.fasterxml.jackson.databind.ObjectMapper;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;

/**
 * Heartbeat client — periodically POSTs to {@code /api/v1/licenses/&#123;id&#125;/heartbeat}.
 *
 * <p>Wire protocol: HMAC-SHA256 of the JSON body using the API key as the secret,
 * delivered in the {@code X-Forge-Signature: sha256=&lt;hex&gt;} header. Matches
 * the Go / Python SDKs byte-for-byte.
 */
public final class HeartbeatClient {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final SecureRandom RNG = new SecureRandom();

    private final String baseUrl;
    private final String apiKey;
    private final String userAgent;
    private final HttpClient http;

    public HeartbeatClient(String baseUrl, String apiKey, String userAgent) {
        this(baseUrl, apiKey, userAgent, HttpClient.newHttpClient());
    }

    public HeartbeatClient(String baseUrl, String apiKey, String userAgent, HttpClient http) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.http = http;
    }

    public Response send(String licenseId, String fingerprint) throws IOException, InterruptedException {
        Map<String, Object> body = new HashMap<>();
        body.put("license_id", licenseId);
        body.put("fingerprint", fingerprint);
        body.put("verifier_version", userAgent);
        body.put("nonce", randomNonceHex());
        body.put("reported_at", Instant.now().toString());

        byte[] raw = MAPPER.writeValueAsBytes(body);
        HttpRequest.Builder req = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/licenses/" + licenseId + "/heartbeat"))
                .header("Content-Type", "application/json")
                .header("X-Forge-API-Key", apiKey)
                .header("X-Forge-Signature", "sha256=" + hmacHex(apiKey, raw))
                .POST(HttpRequest.BodyPublishers.ofByteArray(raw));
        if (userAgent != null && !userAgent.isEmpty()) {
            req.header("User-Agent", userAgent);
        }
        HttpResponse<byte[]> resp = http.send(req.build(), HttpResponse.BodyHandlers.ofByteArray());
        int code = resp.statusCode();
        if (code == 401) throw new IOException("heartbeat: invalid api key (401)");
        if (code == 429) throw new IOException("heartbeat: rate limited (429)");
        if (code >= 500) throw new IOException("heartbeat: server error " + code);
        return MAPPER.readValue(resp.body(), Response.class);
    }

    private static String randomNonceHex() {
        byte[] b = new byte[16];
        RNG.nextBytes(b);
        return HexFormat.of().formatHex(b);
    }

    static String hmacHex(String key, byte[] body) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            return HexFormat.of().formatHex(mac.doFinal(body));
        } catch (Exception e) {
            throw new IllegalStateException("HmacSHA256 unavailable", e);
        }
    }

    public static final class Response {
        public boolean ok;
        public String licenseStatus;
        public boolean multiEnvAnomaly;
        public int nextHeartbeatAfterSeconds;
        public String reason;
    }
}
